### ISSUE TITLE
Persistent State Storage for Improved State Syncing

### ISSUE THREAD

**Issue by bkchr** [paritytech/polkadot-sdk#4]
Currently, during the state syncing process from the network, our node retains its internal state, including downloaded portions of the blockchain state and the target block, in memory. This poses a significant challenge: if the node restarts mid-sync, all the data must be re-downloaded.

To improve this process, I propose persistently storing the internal state outside of memory. This could be in a database or a plain file, allowing us to preserve the internal state across node restarts. This would be especially beneficial when handling a large blockchain state, which doesn't need to remain in memory.

However, one potential issue to consider is if the node is offline for an extended period, the state of the block may no longer be available on full nodes within the network, but only on archive nodes. To mitigate this, we could consider the implementation suggested in [Issue paritytech/substrate#13248](https://github.com/paritytech/substrate/issues/13248).

**Comment by liuchengxu** [comment]
Given that the initial step in https://github.com/paritytech/polkadot-sdk/issues/5053#issuecomment-2370371891 is already underway, as implemented in https://github.com/paritytech/polkadot-sdk/pull/5956, I am planning to initiate work on the persistent state sync improvements.

My immediate focus will be on a pure refactoring of the existing state sync process to modularize and clarify the data processing logic. So that it's super clear what data are retained in memory and how they interact. This refactoring will streamline the addition of persistent state storage and ensure a clear, maintainable transition path for future updates.

**Comment by liuchengxu** [comment]
There is a issue with the original plan described in [this comment](https://github.com/paritytech/polkadot-sdk/issues/5053#issuecomment-2370371891). Specifically, we cannot directly forward the partial state into the database and import it there because resolving the state version requires access to the complete state—or more precisely, the `CODE` key-value. While we could include the state version in the state response, doing so would require modifying the network message. Unfortunately, this is currently blocked due to the lack of network versioning, similar to the other network improvements discussed in issues #5312 and #5397.

To address this limitation, I propose an alternative approach: persisting the received state data to disk, either in a plain file or a database, as suggested in the original idea for this issue. 

With #5956, it's already doable to import the entire state of several GBs into the node. To improve the memory usage further, we can make the state import process from the persistent state data interruptible so that we don't need to hold the entire state in memory. What do you think? @bkchr

**Comment by bkchr** [comment]
@liuchengxu the `StateResponse`s are `CompactProof`s. We should be able to import the nodes from these proofs directly?

**Comment by liuchengxu** [comment]
@bkchr Not really. `StateResponse` being `CompactProof` is false when the mode is `--sync=fast-unsafe`. In addition, `CompactProof` doesn't convey the state version, which is required by the API for importing the state proposed in https://github.com/paritytech/polkadot-sdk/pull/5956/files#diff-f57962eb96265a38bd88c373e76f31d4505d999834e4908461d2950fea2936R610-R615. We have to access to the entire state to read `CODE` storage entry for resolving the state version before importing any state data.

**Comment by bkchr** [comment]
> `StateResponse` being `CompactProof` is false when the mode is `--sync=fast-unsafe`

We could just change it to always request the proofs, but skip to check the proofs for `fast-unsafe`.

> In addition, `CompactProof` doesn't convey the state version, which is required by the API for importing the state proposed in

If we directly import the trie nodes from the proofs, we don't need to reconstruct these nodes as done in the linked pr.

**Comment by liuchengxu** [comment]
>If we directly import the trie nodes from the proofs, we don't need to reconstruct these nodes as done in the linked pr.

 If it could work this way, we can continue the way of forwarding the state response directly to the database. Please review https://github.com/paritytech/polkadot-sdk/pull/5956 and enlighten me there :P
